### DISCUSSION TITLE
Plan for preparing the roadmap

### DISCUSSION THREAD

**Discussion by cgbosse** [torrust/.github#19] [General]
Upon discussing with @josecelano the preparation of the current release planned for end of the month of July 2024 we have concluded the following:

**Until now**
---------------------------------------------

We have essentially done refactoring work improving the code base and essential structure of the entire Torrust project, the purpose was to create a project where the code quality and documentation in addition to the existing two subprojects are easy to build forward and contribute to for new developers given our goal of attracting contributors. So far we have started to have some external contributors who actively picked up some of the easier issues and worked on them.

**Current release**
---------------------------------------------

The current release has taken the code base to a much better level than two years ago. It is now possible for other experience developers knowledgeable of Rust to experience a better onboarding.  I underline the term experienced since from our own team's experience without being knowledgeable of Rust it takes some additional time, since it is not an easy language to learn. 
Other conclusions that we have reached are:

- The tracker is usable as it is, offering the basic required features this type of application has.
- The index is still a wip since it still lacks many basic features, such as a password reset, a proper admin interface, etc.
- Both solutions do not offer anything extraordinary except for being written in Rust with its derived benefits such as security, and resource management.

**Preparation of future roadmap**
---------------------------------------------
Given the above, we need to define a roadmap that makes sense and is aligned with the direction that we decide upon from now onwards.
Things that @josecelano and I think need to be done in preparation for setting new objectives and timelines is to:

1. Research the competition 

There are two types of competition, similar torrent solutions and the IPFS decentralized file share technology.
As for solutions we intend to:
- Index and tracker to see what are the mandatory features to have for both solutions.
- Do some test installations to check these features and performance.
- List their features and define the mandatory features that they offer.
- Look at the possible differentiators that each competitor has.
- Look at current active indexes market or offering (Anime, research, data models, etc.).

As for IPFS:
- We will evaluate if it is worth incorporating it.
- Compare it to Torrents
- Study how it solves different problems of file sharing or problems that we have encountered with torrents so maybe we can learn from it.

2. Decide how we want to differentiate our solutions
There can be many options or areas where we can spend our time on, however; with such a small team we need to find a focus for our efforts. It also needs to be an area which other users and developers do find interesting so that we create value by addressing a niche and also have something that other potential contributors find interesting to contribute to. 

Something that we are considering is reviewing our current feature requests by others and also launching a poll and suggestion board.

Some things that were mentioned in some of our meetings were implementing things such as:
- Web Torrents streaming
- Creating a Torrust client
- Automate categorizing of torrents based on metadata and tagging them in various ways
- Providing core packages focusing on making developing torrent solutions easier
- Creating a "Superapp" based on torrents developing multiple specific purpose plugins

3. Planning
Hereafter, we would then define our goals, set the coming priorities  in the form of new features and set a timeline for our next release.

**Comment by cgbosse** [comment]
- Linked torrents 
- May require a protocol for announcement in tracker for new versioning

**Comment by cgbosse** [comment]
- <https://www.phoronix.com/> - open source news site
- post a main release article on Torrust
-

**Comment by cgbosse** [comment]
After our meeting today I had a call with Jose.
He told me also that he had forgotten about a comment made by the two developers who were reviewing the project with him, Carla and Jesus.

As it seems from what I understood, they mentioned that they were surprised not to see a more developed roadmap since this would help potential contributors know where we are headed since Carla for instance said that she when it comes to contributing it is good to know where a project is headed to see if you want to contribute to it.

This is a topic that @josecelano and I agree that we should look into.  
Making some rough sketches to where we are headed and discuss it  is mandatory, so that already before the release we created this discussion and would like everyone's input, @josecelano and I discussed several things that we thought were relevant but had not input from the rest of the team so far.

**We need to exchange opinions**, and **define the direction** to then **establish the small steps** that will take us there. 

This is necessary not only for the reasons above but also to avoid doing random issues without a specific purpose which can lead to boredom, and also unproductive work and hamper progress long term, especially being a small team. More so however because if we do not it will look as if we have not goal.

@da2ce7:
From past discussions, we think that the ideas that you shared these were interesting but need to be more defined and reflected on a roadmap. Given this @da2ce7, please to share your thoughts on what you would like to do in the coming 12 Months so that we can discuss it, the input from everyone else from the team are also welcome.

@josecelano:
Also please add your comments and suggestions. I believe after talking to you that you also have very relevant insights that we need to consider if we do want to grow our community of users and potential contributors.

@grmbyrn and @mario-nt also feel free to add and contribute to the discussion.

The discussion is also open to externals so that we will share it online.

**Comment by josecelano** [comment]
Hi @cgbosse, regarding long-term goals, I see two directions:

1. Improve Classical Indexes
2. Explore Other Decentralized Alternatives

### Improve Classical Indexes

I think what we are building right now it's another implementation of the "classical" Index.

- BitTorrent client (we don't have a client).
- Torrust Tracker: a centralized tracker.
- Torrust Index: a centralized index linked to the centralized tracker.

In particular, we have been focused on building an easy-to-use simple Index.

I think our solution is good at providing a solution that:

- It consumes few resources. It can be executed in a small VM.
- From the end-user perspective is very easy to use.

I guess our main use case is a public tracker (I prefer to call it Index), where the main goal for the index maintainers is to share their collections. An example could be: https://academictorrents.com/

If we continue working in this direction, I think we have three main t things to improve:

1. Make the **installation easier**. See https://github.com/torrust/torrust-installer
2. Allow users to **customize the Index GUI**, at least the logo and primary colours.
3. **Improve the admin panel**.

I think we are very far from having a fully-featured private tracker. See https://torrust.com/blog/review-and-setup-guide-for-unit3d

### Explore Other Decentralized Alternatives

Some BitTorrent clients include features that act as trackers or GUIs for indexes.

For example, the BitTorrent client has these features:

- Search plugins: you can search for torrents in external indexes.
- RSS: you can subscribe to torrents and even download them automatically.

In most cases, they handle searching for torrents but cannot create and share collections. The only fully decentralized client I know that allows users to create and share content is [Tribler](https://www.tribler.org/). However I think it's very hard to create a new collection (they called channel) automatically from scratch.

I would like to have something like Tribler but with:

- A modern interface
- A bulk import. For example, add a whole dir, automatically create the channel, and include the metadata imported from different formats (embedded in the torrent, external JSON, etc.).
- Private channels (I don't know if that's an existing feature).
- Versioning or updatable torrents/collections. A way to import only changes when there is something new or updated in the collection. This is a fuzzy idea, it may involve the user or be transparent to the user.

**Reply by ldpr** [reply]
> I think we are very far from having a fully-featured private tracker

I think you're being a bit unfair on yourselves by comparing Torrust to full-featured, mature projects like Unit3d and Gazelle. Torrust does a fantastic job at what it does _right now_. There are loads of bells and whistles that can be added sure, but none of them will make the most important thing - downloading a torrent - any faster or easier. If anything, it'd be the opposite. Instead of comparing to the monoliths, to me, Torrust is much better compared to something like nyaa (python: https://github.com/nyaadevs/nyaa or GO: https://github.com/NyaaPantsu/nyaa). In a few years Torrust might surpass nyaa and have gazelle-like features sure, but a lot of those features mightn't be needed by then either. For example, forums, chatboxes, support and even wikis have all largely been replaced by Discord (etc).

**Comment by josecelano** [comment]
Hi @abstralexis,@Aimless321,@alexohneander,@anujkumardarji,@danielroe,@ftsimas,@GGLinnk,@hungfnt,@hungfnt,@ldpr,@makefu,@mario-nt,@mickvandijke,@naim94a,@pataquets,@pcarles,@postmeback,@Power2All,@ShrirangB,@si14,@Wolfremium13,@Zorlin

First of all, I want to thank all of you for your past contributions to Torrust. Your efforts have been instrumental in making this project what it is today, and we truly appreciate the time, skill, and dedication you've brought to the community.

As we plan the roadmap for Torrust in 2025, we want to hear from you! Your insights and ideas are incredibly valuable in shaping the future of the project. If you have suggestions, feature requests, or thoughts on the direction we should take, I’d love for you to join this discussion and share your perspective.  

Let’s collaborate to make Torrust even better in 2025. Looking forward to your input!  

Thank you again for everything you've contributed to the project so far!

**Reply by Zorlin** [reply]
Thanks folks Torrust is amazing.

**Reply by ldpr** [reply]
Torrust has really taken on a life of it's own with all your work Jose, it's very impressive and you (and the team) deserve all the recognition & plaudits coming your way.
Regarding the future, I'd _love_ for torrust to remain (optionally) simple and ideally, to allow more  customization. We did a bunch of testing with 3.0 and have some feedback but I want to chat with some of the guys first before forwarding it on here.

I believe user-roles are already in the pipeline. Additionally RSS would be amazing (a lot of our users are asking about this lately).
Otherwise though, thanks for everything and a very belated congrats on the 3.0 release :D

**Reply by josecelano** [reply]
Thank you @Zorlin @ldpr. Yesterday, we had a meeting to discuss the roadmap. Today, I will start working on collecting and organizing all the feature requests together with our ideas. We will also prioritise them and create public milestones so the community can know precisely what to expect from the project in the next 1 or 2 years.

My personal preferences are:

- All projects
  - V2 support
  - Other BEPs implementations
  - Fix bugs, keep the dependencies updated, docs, and code clean.
- Tracker+Index
  - Installer
- Tracker   
  - Webtorrent
  - Improve statistics
- Index
  - Improve search
  - Improve importation: exporter/importer
  - RSS
- Custom client?
  - Or maybe a plugin for another client. I think bulk upload is tedious. This could be related to the export/import feature

However, we know there are other essential missing features, especially for index admins:

- UX/UI customizations
- Backend administration
